ostd/src/mm/io.rs

```diff
@@ -536,9 +493,39 @@ impl<'a, Space> VmReader<'a, Space> {
         unsafe { self.cursor = self.cursor.add(nbytes) };
         self
     }
+
+    fn read_fallible_general<WF>(
+        &mut self,
+        writer: &mut VmWriter<'_, WF>,
+    ) -> core::result::Result<usize, (Error, usize)> {
+        let copy_len = self.remain().min(writer.avail());
+        if copy_len == 0 {
+            return Ok(0);
+        }
+
+        // SAFETY: This method is only implemented when one of the operated
+        // `VmReader` or `VmWriter` is in user space.
+        // The the corresponding page table of the user space memory is
+        // guaranteed to be activated due to its construction requirement.
+        // The kernel space memory range will be valid since `copy_len` is the minimum
+        // of the reader's remaining data and the writer's available space, and will
+        // not overlap with user space memory range in physical address level if it
+        // represents typed memory.
+        let copied_len = unsafe {
+            let copied_len = memcpy_fallible(writer.cursor, self.cursor, copy_len);
+            self.cursor = self.cursor.add(copied_len);
+            writer.cursor = writer.cursor.add(copied_len);
+            copied_len
+        };
+        if copied_len < copy_len {
+            Err((Error::PageFault, copied_len))
+        } else {
+            Ok(copied_len)
+        }
+    }
 }
 
-impl<'a> From<&'a [u8]> for VmReader<'a> {
+impl<'a, Fallibility> From<&'a [u8]> for VmReader<'a, Fallibility> {
     fn from(slice: &'a [u8]) -> Self {
```

Comment:
Only implementing `From<&'a [u8]>` for `VmReader<'a, NonFallable>` is better. Converting from `VmReader<'a, NonFallable>` to `VmReader<'a, Fallable>` is kind of a lossy one. Thus, the API design should discourage such usage by asking the developer to do it explicitly.

```rust
impl<'a> VmReader<'a, NonFallable> {
    pub fn to_fallable(self) -> VmReader<'a, Fallable> {
        todo!()
    }
}
```

So the user code would convert a slice to a fallable in two steps.

```rust
let buf = Vec::new();
let reader: Reader<'_, Fallable> = buf
    .as_slice()
    .into() // Rust compiler should be smart enough to infer this type, although I am not 100% sure.
    .to_fallable();
```